flink-clients/src/main/java/org/apache/flink/client/deployment/application/ApplicationDispatcherBootstrap.java

```diff
@@ -147,35 +146,34 @@ public void stop() {
-                                            t, UnsuccessfulExecutionException.class);
-                            if (maybeException.isPresent()) {
-                                final ApplicationStatus applicationStatus =
-                                        maybeException.get().getStatus();
-                                if (applicationStatus == ApplicationStatus.CANCELED
-                                        || applicationStatus == ApplicationStatus.FAILED) {
-                                    LOG.info("Application {}: ", applicationStatus, t);
-                                    return dispatcherGateway.shutDownCluster(applicationStatus);
-                                }
-                            }
-
-                            LOG.warn("Application failed unexpectedly: ", t);
-                            this.errorHandler.onFatalError(
-                                    new FlinkException("Application failed unexpectedly.", t));
-
-                            return FutureUtils.<Acknowledge>completedExceptionally(t);
-                        })
-                .thenCompose(Function.identity());
+        final CompletableFuture<Acknowledge> shutdownFuture =
+                applicationCompletionFuture
+                        .handle(
+                                (ignored, t) -> {
+                                    if (t == null) {
+                                        LOG.info("Application completed SUCCESSFULLY");
+                                        return dispatcherGateway.shutDownCluster(
+                                                ApplicationStatus.SUCCEEDED);
+                                    }
+                                    final Optional<UnsuccessfulExecutionException> maybeException =
+                                            ExceptionUtils.findThrowable(
+                                                    t, UnsuccessfulExecutionException.class);
+                                    if (maybeException.isPresent()) {
+                                        final ApplicationStatus applicationStatus =
+                                                maybeException.get().getStatus();
+                                        if (applicationStatus == ApplicationStatus.CANCELED
+                                                || applicationStatus == ApplicationStatus.FAILED) {
+                                            LOG.info("Application {}: ", applicationStatus, t);
+                                            return dispatcherGateway.shutDownCluster(
+                                                    applicationStatus);
+                                        }
+                                    }
+                                    LOG.warn("Application failed unexpectedly: ", t);
+                                    return dispatcherGateway.shutDownClusterExceptionally(t);
+                                })
```

Comment:
I think this does not really solve the described problem. If the `Dispatcher` loses leadership, then it will call `ApplicationDispatcherBootstrap.stop` which will cancel the `applicationCompletionFuture`. In this case, I think we should neither fail nor shut down the cluster because we only want to stop the `ApplicationDispatcherBootstrap` and maybe restart it once we have regained the leadership.

Reply:
Hmm, I think I may have misinterpreted the issue description then 🤔 I initially thought that don't want any exceptions (even from the user code) to shut down the process fatally.